### PR TITLE
pkp/pkp-lib#5642 Implemented user report with a CSV serializer.

### PR DESCRIPTION
**Pull request by jonasraoni**
Notes:
- I've tried to make the code modular, so it's possible to add/remove columns, tweak the query builder to add more data and I've left it ready to add more output formats.

- I've added the `STRING_AGG`/`GROUP_CONCAT` to avoid sacrificing the database performance with extra queries. These functions will require a bump in the base MySQL version... In fact, I would vote to spread such optimization to other parts of the system, the "JSON" variants of these functions would fit perfectly (while also requiring a higher bump in the versions) =]

- I had a version of this code with an extra filter to select the journals (instead of just exporting users from the current journal), if that's useful, I can bring it back and tweak the output (add a row to group some columns by journal).

- I was going to add more filters (filter by name, registration date, etc), but I'm not sure it's needed.

- Check if it's too verbose the way I coded my `QueryListener` (here's a less verbose one from a removed commit: https://github.com/jonasraoni/pkp-lib/commit/37c93c84792b80657f51772f5ef6b7b0b934a18f): Well, I've added an interface to make the callback params more "typed" and also a `QueryOptions` to avoid using a generic `$args` array.

- I've somehow [hacked the `<pkp-form>` ](https://github.com/pkp/pkp-lib/compare/master...jonasraoni:feature/master/i5642-add-user-export?expand=1#diff-0870af1111a54a09ea354541157712bbR41) to avoid creating a customized Vue container (or adding such feature to the Form.vue itself), so the back-end is just sending a redirection URL to the `api/users/report`.

- Should I check how to add tests for this feature? I didn't check yet how the tests are structured and how to run them locally 👀

- I was going to extend the Illuminate query, in order to remove the custom statements for MySQL/PostgreSQL, but I didn't see a built-in way to do it, besides the `Macroable` trait (which wasn't present where I needed)... If someone knows it, great! Otherwise I can research a bit if there's a not very cursed way to do it (or just desist and create a `CapsuleExtensions`).

### CODE REVIEW

**Comment by NateWr** [comment]
Thanks @jonasraoni! I've just taken a quick look, but I'll do a proper code review soon. At first glance, I am nervous about the complex architecture required to implement this. Did you explore adding the `userGroupIds` filter to the existing `UserQueryBuilder`, and simply using the existing service class method like this:

```
$usersIterator = Services::get('user')->getMany([
  'contextId' => 1,
  'userGroupIds' => [1,2],
]);
```

That should bring with it all of the user data, including orcid, biography, etc, and will ensure that all of the appropriate hooks are hit so that plugins extending the user data for one part of the software will also get those extensions applied in the user export.

You probably already considered this and found it insufficient. Could you explain why so I can get my head around it?

**Comment by jonasraoni** [comment]
Great! I should check it as well, this is one of the notes that I forgot to add \o/

Yeah, I thought about incrementing the `PKPUserQueryBuilder`, then I wanted to raise some questions and decided to leave the discussion for another time.

- My intention was to add some extra data to the result set, due to performance (right now I'm loading just the disabled notifications + list of user groups, but such possibility will be useful for the PS customizations), while keeping the possibility of applying filters. So using the `PKPQueryBuilder` directly was the way I found to achieve it (as I've extended the `PKPUserQueryBuilder`, looks like the same hooks will be called).
- The `PKPUserService` will issue an extra query under the hood to load the user settings. Maybe we can replace this mechanism by [these aggregator functions](https://dev.mysql.com/doc/refman/5.7/en/aggregate-functions.html#function_json-objectagg)? It would require some tweaks to accommodate the locale and type (maybe `JSON_OBJECTAGG(CONCAT(setting_name, '_', locale), CONCAT(setting_type, ':', setting_value))`?!), but it's an acceptable trade-off for me 😁
- It's nice to have a single endpoint to load/modify data, but I wasn't sure it's ok to bloat the `getQuery` with every little thing I might need, so that's why I've decided to extend instead of increment, although I think the `userGroupIds` is ok to be added there...
There was something more I wanted to comment, but I can't remember 🤔

File: api/v1/users/PKPUserHandler.inc.php
Comment (jonasraoni) [inline api/v1/users/PKPUserHandler.inc.php:41]: See #6142

File: api/v1/users/PKPUserHandler.inc.php
Comment (jonasraoni) [inline api/v1/users/PKPUserHandler.inc.php:27]: As these roles already have access to all users, I guess it's ok to keep the same for the report.

File: api/v1/users/PKPUserHandler.inc.php
Comment (jonasraoni) [inline api/v1/users/PKPUserHandler.inc.php:307]: Just left it here in case there's interest to add other formats, if not, I can remove the `abstraction` =]

File: classes/services/PKPUserService.inc.php
Comment (jonasraoni) [inline classes/services/PKPUserService.inc.php:622]: Hmm, I see this shouldn't have a default value.

File: classes/services/PKPUserService.inc.php
Comment (jonasraoni) [inline classes/services/PKPUserService.inc.php:640]: An opportunity to customize the report will be needed for the PS team =]

File: classes/user/report/QueryBuilder.inc.php
Comment (jonasraoni) [inline classes/user/report/QueryBuilder.inc.php:23]: I'd like to move most of the code here to the `PKPUserQueryBuilder`, but instead of many left joins, I'd like to use what I've proposed in the beginning of this discussion.

File: classes/user/report/Report.inc.php
Comment (jonasraoni) [inline classes/user/report/Report.inc.php:23]: Extended the \IteratorAggregate just to use `foreach($report as $dataRow)`

File: classes/user/report/mappings/Notifications.inc.php
Comment (jonasraoni) [inline classes/user/report/mappings/Notifications.inc.php:100]: As the `NotificationSettingsForm::getNotificationSettingsMap` is a protected method, I've created this anonymous class just to extract it 👀
Alternatives:
- Add the constants manually (will require an extension to get the ones defined at the application level)
- Make the method public
- etc.

File: templates/stats/users.tpl
Comment (jonasraoni) [inline templates/stats/users.tpl:41]: When submitted, the backend will merely return as response an URL to the `api/users/report` with the submitted parameters.

File: classes/user/report/serializer/CSV.inc.php
Comment (jonasraoni) [inline classes/user/report/serializer/CSV.inc.php:52]: ⛔

File: classes/user/report/Report.inc.php
Comment (jonasraoni) [inline classes/user/report/Report.inc.php:41]: Just wrote this to have a working report by default.
I've created these `Mappings\*` in the end, just to organize a bit the fields... So they are kind of disconnected from the implementation, if needed I can add an `addUserGroups()/addMappingGroup($mapping)`...

File: classes/user/report/mappings/Notifications.inc.php
Comment (jonasraoni) [inline classes/user/report/mappings/Notifications.inc.php:63]: Is there something ready to add such exceptions in the code base? If not I'll probably create a `PKPCapsule extends Manager` to insert such custom methods, in case nobody has a better idea...

**Comment by NateWr** [comment]
Thanks @jonasraoni! I understand your concerns about performance. The current architecture, where additional data is loaded per row in the `_fromRow()` method, is not the most performant approach. I know that Alec has raised this as well and is considering some alternatives to improve this.

That said, I want to avoid introducing one-off abstractions into the codebase. Because of the size of our code base, and the diversity of developers who will interact with it over a long period, there's a high cost to introducing abstractions to solve isolated needs. That's because each abstraction requires new learning from a developer and new documentation and training materials to be created for it. If we re-use the same architecture when creating user exports a developer can bring their existing knowledge to this part of the codebase.

If we are going to solve the performance problems of attaching related objects during database retrieval, we need to do it in a way that will be rolled out across all of our codebase, and in a way that takes into account additional needs like documentation and third-party extensions. I'd love to link you up with Alec directly to discuss this problem and figure out an approachable abstraction that works across our codebase. For example, we plan to merge our DAO and QueryBuilder classes.

That said, I'm concerned about merging it in here without going through that process and making sure it's used consistently throughout our application. 

In this case, I think that the overall database queries are not the right bottleneck to optimise away. Instead, our bottleneck is going to be the size of the overall users table and the need to build the CSV file in chunks. From the user's perspective, we can deliver the CSV file 5, 10, 30 minutes later. But hammering the server with a single request that returns a CSV with 100k rows is where problems are going to arise.

We're planning to work on a task schedule, so that's not a bottleneck that you need to solve in this pull request. Still, I'd rather see us use the existing architecture, knowing it's not the most performant, and hold fire on introducing new architecture to solve a problem that is best solved elsewhere.

To address performance, you could build the REST API endpoint to work in batches. So the API response would send status information like 350/12031 users exported, and in the end provide a link to the download the report. We can then easily integrate this with a task runner when we get one.

In terms of PS's need to attach additional data, I think that this can be done with the existing architecture. When handling the request, you can define what properties that should be returned for each row when converting the object to a JSON object. Instead of calling `getSummaryProperties` or `getBackendProperties`, the report endpoint could call `getProperties` directly with the property structure it wants for the CSV files. The `User::getProperties::values` hook can be used to configure properties not configured by the base service class.

I appreciate that this is a really frustrating way to reach this point, as you've done a lot of good work devising performance enhancements. I don't want to lose that value if we can, but I do want to find a way to reduce the number of abstractions that we're using. Let's try to have a call this week and see if we can talk through some of these details.

**Comment by jonasraoni** [comment]
@NateWr That was a long text! I understand your concerns :)

Specialized vs generic code is complex, would be great to keep them together, but sometimes it's tough. My opinion is that for general cases, the latter is great (less maintenance), but the former shouldn't be entirely discarded. As the project is in a kind of intermediate state, with old and new things mixed up, I think it's a good time to raise such questions, but this small task isn't really the right place for it =]

With that being said, it's not a big deal for me to revert some changes (previously it was using the DAOs to load the notifications).

Yeah, breaking such tasks in batches is a kind of requirement. I didn't see such user-interface with progress elsewhere, but it came to my mind, although I ended up forgetting to cook it up.

**Comment by NateWr** [comment]
Thanks Jonas! I spoke to Alec yesterday about your optimisation ideas. He's got a lot of catching up to do after his break but I think the two of you should get together and talk through ways to reduce the number of queries.

So for this let's move the `userGroupIds` filter into the `UserQueryBuilder`, and add it to the `PKPUserService::getQueryBuilder()` method so that we can reuse it when we want. We can then support that query param in the `/users` API endpoint fairly easily.

In terms of adding data onto the report response, I'm open on how you go about doing it -- whether you want to write a new SQL query or extend the `getProperties` method with additional properties. But I'd prefer if this part of it went into a method in `PKPUserService`. Even if it's just a wrapper method around logic, our goal with the Service class is to provide a discoverable API that can be used to fulfil http and cli requests.

**Comment by asmecher** [comment]
Thanks, @jonasraoni and @NateWr, this is really interesting work!

Long-term we have such a problem with code expanding beyond our ability to maintain it that I'd like to take a restricted approach to cases where there's an optimization vs. code size trade-off -- we should go with the shorter code unless we have a real problem with performance. (And ideally look for ways to facilitate both goals at once, which is often possible.) The user export will be so rarely used that even if it lit the server on fire for a minute I doubt anyone would notice. So in this case I'd definitely prefer less abstraction.

That said: @jonasraoni, Nate and I have been working through some proposals for general improvements to the DAOs and QueryBuilders and I think you'd be a huge help in refining those and trying out some of the approaches. Some of your ideas here might have more impact there. If you're interested, we can set up a call to go over it?

**Comment by jonasraoni** [comment]
@asmecher Due to my limited working time, I don't have a deep knowledge about the code yet, but feel free to share the proposals/include me in the discussion, I have other subjects to discuss =]

**Comment by asmecher** [comment]
Followed up in Slack!

**Comment by jonasraoni** [comment]
Ah, just leaving a note that I've removed my customizations =]

**Comment by asmecher** [comment]
Thanks, Jonas!

I think the `Mapping` abstraction needs to be considered/introduced beyond this one report for it to justify the LOC overhead... essentially it is used for 3 use cases (User Groups, Notifications, and normal columns). I don't think Notifications will be a useful column for this report (I suggest removing it, at least until the notifications system is overhauled). That leaves the normal case, which is kind of a no-op, and User Groups, which can be essentially a `join(...)` call.

We'll probably need to work on a mapping abstraction -- in particular to bolster the API, but it could conceivably be used for reports too -- but I think that should drive the development rather than this report. I think you'll probably lose a few CLOCs in the process (and not add the technical debt of needing to refactor the other reports to use it).

**Comment by jonasraoni** [comment]
@asmecher I've ignored the "which can be essentially a join(...) call"... Please elaborate, before I take the wrong path :)
By now I've just removed the notifications (it was embedded in my original plugin) and inlined/removed the abstractions into a single file.

File: api/v1/users/PKPUserHandler.inc.php
Comment (asmecher) [inline api/v1/users/PKPUserHandler.inc.php:330]: Is it possible to use `fopen('php://stdout', 'w')` or something to avoid a temporary file? The only challenge woudl be getting the headers to be sent first, the way Smarty's response chaining works.

File: classes/components/forms/statistics/users/ReportForm.inc.php
Comment (asmecher) [inline classes/components/forms/statistics/users/ReportForm.inc.php:33]: Off the top of my head I think you could avoid `iterator_to_array($something->toIterator())` by just calling `$something->toArray()`.

File: classes/components/forms/statistics/users/ReportForm.inc.php
Comment (asmecher) [inline classes/components/forms/statistics/users/ReportForm.inc.php:30]: We're gradually shifting from `array(...)` syntax to `[...]` -- feel free to use the more concise syntax for new code!

File: classes/components/forms/statistics/users/ReportForm.inc.php
Comment (asmecher) [inline classes/components/forms/statistics/users/ReportForm.inc.php:24]: We're currently using inline `{` syntax for function declarations.

File: classes/services/queryBuilders/PKPUserQueryBuilder.inc.php
Comment (asmecher) [inline classes/services/queryBuilders/PKPUserQueryBuilder.inc.php:164]: I'm bikeshedding here, but a more concise way of doing this would be an explicit `(array)` typecast.

https://www.php.net/manual/en/language.types.array.php#language.types.array.casting:
> In other words, (array)$scalarValue is exactly the same as array($scalarValue).

If `null` values are allowed, the self-documentation should reflect that. (Note that `(array) null === []` so an explicit cast might be useful there too.)

File: classes/services/queryBuilders/PKPUserQueryBuilder.inc.php
Comment (asmecher) [inline classes/services/queryBuilders/PKPUserQueryBuilder.inc.php:434]: If we move to initializing `$this->columns` here, we may as well make it a local variable and not a class member. @NateWr, is this a pattern that's useful elsewhere?

File: classes/services/queryBuilders/PKPUserQueryBuilder.inc.php
Comment (asmecher) [inline classes/services/queryBuilders/PKPUserQueryBuilder.inc.php:476]: (If you move away from null `userGroupIds` towards empty arrays, as I mentioned above, you'd need to adjust this to `!empty` -- but again, bikeshedding.)

File: classes/user/Report.inc.php
Comment (asmecher) [inline classes/user/Report.inc.php:87]: I'm not sure I'm reading it right, but it looks to me like by escaping the HTML here, the CSV will include double-escaped content. I think it would be better to escape where the data is presented -- `fputcsv` should take care of CSV's escaping needs, and HTML escaping should happen in the Smarty template (https://github.com/pkp/pkp-lib/pull/6240/files#diff-5bcedff5ad35212dc345cf262eb439cd4e78e509d4032bc8d6f0fd27f9978f2bR32).

File: classes/user/Report.inc.php
Comment (asmecher) [inline classes/user/Report.inc.php:7]: I'm still thinking this abstraction is in the wrong place or at the wrong time. The only current calling code is the user report, where a mapping via
```
switch ($columnName) {
    case 'family_name': return $user->getFamilyName($locale);
    ...
}
```
...would take care of this concisely. If it's an abstraction that is intended for use elsewhere, I'd rather not introduce it until it's been generalized (out of `\PKP\User\Report`) and called from a couple of places to establish a new pattern. Adding abstractions with single calling cases will lead to a profusion of these, which I'm sensitive to as we have a long history of that :)

For now I'd recommend sticking with a more compact implementation, and bringing the mapping stuff to the optimization group for some experimentation beyond the context of the user report -- I think @NateWr is interested in this for the API, for example.

**Review by asmecher — CHANGES_REQUESTED** [review]
Thanks, Jonas, and with apologies for the delay in responding to this, I do have a few more changes to suggest -- in particular slimming back the mapping stuff even further and bringing it to the optimization channel for some broader discussion. It's a good direction to be heading, but needs to be introduced more broadly once we've hashed it over there.

File: classes/services/queryBuilders/PKPUserQueryBuilder.inc.php
Comment (NateWr) [reply classes/services/queryBuilders/PKPUserQueryBuilder.inc.php:434]: I believe it's necessary for the hook, which receives `$this`. That's why the `select()` method is not called until after the hook, so that we can modify the fetched columns.

File: classes/user/Report.inc.php
Comment (NateWr) [reply classes/user/Report.inc.php:7]: Yes, the mappings overlap a lot with the `UserService::getProperties()` method. For now I'd suggest extending `getProperties` as needed. And in the future we can convert the `getProperties` pattern to a class-based mapping, similar to how you've done it here, and roll that out to all of the entities.

Comment (jonasraoni) [reply classes/user/Report.inc.php:7]: @asmecher FYI I've ported this code from a plugin, and we had the need to extend it with some custom user data (e.g. the enabled notifications + custom fields), I'll take a look in the comments later. But I think it's not a big deal, as we can still use the plugin :)

Comment (NateWr) [reply classes/user/Report.inc.php:7]: You should be able to pass into `getProperties` the set of properties you want, and hook into the method to provide values for any properties that are not handled in the core application.

Comment (asmecher) [reply classes/user/Report.inc.php:7]: @jonasraoni, just checking -- is this still underway? We're hoping to wrap up the release candidate tomorrow.

File: classes/components/forms/statistics/users/ReportForm.inc.php
Comment (jonasraoni) [reply classes/components/forms/statistics/users/ReportForm.inc.php:33]: I'll check =]

File: api/v1/users/PKPUserHandler.inc.php
Comment (jonasraoni) [reply api/v1/users/PKPUserHandler.inc.php:330]: Yeah, I had an intention here (related to blocking the script under slow connections), but it was based in an old myth from my head, I've just killed it a test =]

File: classes/components/forms/statistics/users/ReportForm.inc.php
Comment (jonasraoni) [reply classes/components/forms/statistics/users/ReportForm.inc.php:30]: This one escaped from a ctrl+c/v :D

File: classes/components/forms/statistics/users/ReportForm.inc.php
Comment (jonasraoni) [reply classes/components/forms/statistics/users/ReportForm.inc.php:24]: Ok! I've just spotted the current style when there's return types:
```
function test() : void {
}
```

File: classes/services/queryBuilders/PKPUserQueryBuilder.inc.php
Comment (jonasraoni) [reply classes/services/queryBuilders/PKPUserQueryBuilder.inc.php:434]: I've modified this part due to something that it's already removed, but I think it makes more sense to keep it this way :)

File: classes/user/Report.inc.php
Comment (jonasraoni) [reply classes/user/Report.inc.php:87]: My idea was to remove any HTML/entity from the output, so the CSV can be easily imported/consumed into another system (e.g. email delivery service). I'm not sure the fields I selected may have HTML content, I just wanted to output safely.

File: classes/user/Report.inc.php
Comment (jonasraoni) [reply classes/user/Report.inc.php:7]: @asmecher 😵 I've been busy with other things, but I'll give it some time today!

File: classes/services/queryBuilders/PKPUserQueryBuilder.inc.php
Comment (jonasraoni) [reply classes/services/queryBuilders/PKPUserQueryBuilder.inc.php:164]: Hmm, this method has been introduced 2 days ago by another PR, so I'll just merge it.

File: classes/user/Report.inc.php
Comment (asmecher) [reply classes/user/Report.inc.php:87]: It's a good idea, but should be targeted to only columns that might contain HTML; if you're not sure about any, post here and I can confirm.

File: classes/components/forms/statistics/users/ReportForm.inc.php
Comment (jonasraoni) [reply classes/components/forms/statistics/users/ReportForm.inc.php:33]: This was valid, until few commits ago, I've just rebased and it's lost :D

File: classes/user/Report.inc.php
Comment (jonasraoni) [reply classes/user/Report.inc.php:7]: I've totally removed the mappings, now there's a `_getHeadings` and a `_getDataRow` methods, I could join them in one method, which would look somehow more organized (`return [__('common.id') => $user->getId(), etc];`), but it would be a bit sad to use array_keys/array_values to fill the headings/rows 🤔

File: classes/user/Report.inc.php
Comment (jonasraoni) [reply classes/user/Report.inc.php:46]: I took a look and I think none of them will have HTML, but it was a safe assumption to support my unknown mappings :)
If you want to review, [here are the fields](https://github.com/pkp/pkp-lib/pull/6240/files#diff-752409eb9449c35619d76b53814be94664528e44146454fcd740af8b84f3aa7eR95).

Comment (asmecher) [reply classes/user/Report.inc.php:46]: Yes, none of those support HTML -- thanks for checking!

File: api/v1/users/PKPUserHandler.inc.php
Comment (jonasraoni) [reply api/v1/users/PKPUserHandler.inc.php]: I was taking a look at the Slim implementation and they are using `php://temp` to buffer the output, so I've just replaced by it, with some good luck, the file system won't be used...

Also, I've tried to skip the Slim conventions, by just writing to the stdout and exiting, but looks it's using the ob_* functions. I'll check how to disable the buffering 🤔

File: classes/user/Report.inc.php
Comment (jonasraoni) [reply classes/user/Report.inc.php]: Ok! I've removed the escaping.

**Comment by asmecher** [comment]
Hi @jonasraoni! Is this ready for a last review? (Not sure if you're still working on part of it.)

**Comment by jonasraoni** [comment]
@asmecher yeah, I've already stopped touching it. There are two "conversations" waiting for comments =]

**Comment by asmecher** [comment]
Thanks, @jonasraoni! I see https://github.com/pkp/pkp-lib/pull/6240/files#r531726256 but I think I might've missed the other.

**Comment by jonasraoni** [comment]
@asmecher The other is from this conversation: https://github.com/pkp/pkp-lib/pull/6240#discussion_r528001219

The code became: https://github.com/pkp/pkp-lib/pull/6240/files#diff-9893b5280e6e37e66f44b73222cd375d4874769cc052e2449b09df2cb72c0cf9R320-R326

File: api/v1/users/PKPUserHandler.inc.php
Comment (asmecher) [reply api/v1/users/PKPUserHandler.inc.php]: OK, thanks! That works for now, but if/when either of us get the chance it would be good to tinker with ways to let both coexist without violence.

**Comment by asmecher** [comment]
Merged -- thanks, @jonasraoni, especially for sticking with it through all the discussion!

File: api/v1/users/PKPUserHandler.inc.php
Comment (jonasraoni) [reply api/v1/users/PKPUserHandler.inc.php]: @asmecher Oh, I was waiting for a comment before the merge :)

For example, I think the code below would work, due to the Slim buffer, but it would end up writing to a temp file indirectly:

```php
$report->serialize(fopen('php://output', 'w+'));
return $response
	->withHeader('content-type', 'text/comma-separated-values')
	->withHeader('content-disposition', 'attachment; filename="user-report-' . date('Y-m-d') . '.csv"');
```

**Pull request merged**
